snapshot_create.go

```diff
@@ -0,0 +1,179 @@
+	s.waitForCompletion = &waitForCompletion
+	return s
+}
+
+// Pretty indicates that the JSON response be indented and human readable.
+func (s *SnapshotCreateService) Pretty(pretty bool) *SnapshotCreateService {
+	s.pretty = pretty
+	return s
+}
+
+// BodyJson is documented as: The snapshot definition.
+func (s *SnapshotCreateService) BodyJson(body interface{}) *SnapshotCreateService {
+	s.bodyJson = body
+	return s
+}
+
+// BodyString is documented as: The snapshot definition.
+func (s *SnapshotCreateService) BodyString(body string) *SnapshotCreateService {
+	s.bodyString = body
+	return s
+}
+
+// buildURL builds the URL for the operation.
+func (s *SnapshotCreateService) buildURL() (string, url.Values, error) {
+	// Build URL
+	path, err := uritemplates.Expand("/_snapshot/{repository}/{snapshot}", map[string]string{
+		"snapshot":   s.snapshot,
+		"repository": s.repository,
+	})
+	if err != nil {
+		return "", url.Values{}, err
+	}
+
+	// Add query string parameters
+	params := url.Values{}
+	if s.pretty {
+		params.Set("pretty", "1")
+	}
+	if s.masterTimeout != "" {
+		params.Set("master_timeout", s.masterTimeout)
+	}
+	if s.waitForCompletion != nil {
+		params.Set("wait_for_completion", fmt.Sprintf("%v", *s.waitForCompletion))
+	}
+	return path, params, nil
+}
+
+// Validate checks if the operation is valid.
+func (s *SnapshotCreateService) Validate() error {
+	var invalid []string
+	if s.repository == "" {
+		invalid = append(invalid, "Repository")
+	}
+	if s.snapshot == "" {
+		invalid = append(invalid, "Snapshot")
+	}
+	if len(invalid) > 0 {
+		return fmt.Errorf("missing required fields: %v", invalid)
+	}
+	return nil
+}
+
+// Do executes the operation.
+func (s *SnapshotCreateService) Do(ctx context.Context) (*SnapshotCreateResponse, error) {
+	// Check pre-conditions
+	if err := s.Validate(); err != nil {
+		return nil, err
+	}
+
+	// Get URL for request
+	path, params, err := s.buildURL()
+	if err != nil {
+		return nil, err
+	}
+
+	// Setup HTTP request body
+	var body interface{}
+	if s.bodyJson != nil {
+		body = s.bodyJson
+	} else {
+		body = s.bodyString
+	}
+
+	// Get HTTP response
+	res, err := s.client.PerformRequest(ctx, "PUT", path, params, body)
+	if err != nil {
+		return nil, err
+	}
+
+	// Return operation response
+	ret := new(SnapshotCreateResponse)
+	if err := json.Unmarshal(res.Body, ret); err != nil {
+		return nil, err
+	}
+	return ret, nil
+}
+
+// SnapshotCreateResponse is the response of SnapshotCreateService.Do.
+type SnapshotCreateResponse struct {
+	// Accepted indicates whether the request was accepted by elasticsearch.
+	// It's available when waitForCompletion is false.
+	Accepted *bool `json:"accepted"`
+
+	// Snapshot is available when waitForCompletion is true.
+	Snapshot *struct {
+		Snapshot          string        `json:"snapshot"`
+		UUID              string        `json:"uuid"`
+		VersionID         int           `json:"version_id"`
```

Comment:
Did you compare these fields with what ES really sends, or did you make an educated guess? I typically consult the Java source to see what types ES really sends and more often than not, I find fields that are undocumented but sent anyway.